File: packages/shared-types/src/scene.ts
import { z } from 'zod';

export const SceneTypeSchema = z.enum([
  'deterministic_animation',
  'generated_still_with_motion',
  'veo_cinematic',
  'code_trace',
  'system_design_graph',
  'summary_scene',
]);

export const SceneStatusSchema = z.enum(['pending', 'generating', 'rendered', 'error']);

export const VisualElementSchema = z.object({
  type: z.string().default(''),
  description: z.string().default(''),
  position: z.string().default(''),
  style: z.string().default(''),
});

export const AnimationBeatSchema = z.object({
  timestamp_sec: z.number().default(0),
  action: z.string().default(''),
  description: z.string().default(''),
});

export const AssetRequestSchema = z.object({
  type: z.string().default(''),
  prompt: z.string().default(''),
  provider: z.string().default(''),
});

export const SceneSpecSchema = z.object({
  scene_id: z.string().default(''),
  title: z.string().default(''),
  learning_objective: z.string().default(''),
  source_refs: z.array(z.string()).default([]),
  scene_type: z.string().default('deterministic_animation'),
  render_strategy: z.string().default('default'),
  duration_sec: z.number().default(30),
  narration_text: z.string().default(''),
  on_screen_text: z.array(z.string()).default([]),
  visual_elements: z.array(VisualElementSchema).default([]),
  animation_beats: z.array(AnimationBeatSchema).default([]),
  asset_requests: z.array(AssetRequestSchema).default([]),
  veo_prompt: z.string().nullable().optional(),
  image_prompt: z.string().nullable().optional(),
  music_mood: z.string().default('neutral'),
  validation_notes: z.string().default(''),
});

export const SceneSchema = z.object({
  id: z.string().uuid(),
  lesson_id: z.string().uuid(),
  scene_order: z.number().int(),
  scene_type: SceneTypeSchema,
  title: z.string(),
  duration_sec: z.number(),
  render_strategy: z.string(),
  source_refs_json: z.array(z.unknown()).nullable().optional(),
  narration_text: z.string().nullable().optional(),
  on_screen_text_json: z.union([z.array(z.unknown()), z.record(z.unknown())]).nullable().optional(),
  scene_spec_json: SceneSpecSchema.nullable().optional(),
  status: SceneStatusSchema,
  created_at: z.string(),
  updated_at: z.string().nullable().optional(),
});

export type SceneType = z.infer<typeof SceneTypeSchema>;
export type SceneStatus = z.infer<typeof SceneStatusSchema>;
export type VisualElement = z.infer<typeof VisualElementSchema>;
export type AnimationBeat = z.infer<typeof AnimationBeatSchema>;
export type AssetRequest = z.infer<typeof AssetRequestSchema>;
export type SceneSpec = z.infer<typeof SceneSpecSchema>;
export type Scene = z.infer<typeof SceneSchema>;

File: apps/api/app/models/scene.py
import enum
import uuid
from datetime import datetime, timezone

from sqlalchemy import Column, DateTime, Enum, Float, ForeignKey, Integer, JSON, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.db.base import Base


class SceneType(str, enum.Enum):
    deterministic_animation = "deterministic_animation"
    generated_still_with_motion = "generated_still_with_motion"
    veo_cinematic = "veo_cinematic"
    code_trace = "code_trace"
    system_design_graph = "system_design_graph"
    summary_scene = "summary_scene"


class SceneStatus(str, enum.Enum):
    pending = "pending"
    generating = "generating"
    rendered = "rendered"
    error = "error"


class Scene(Base):
    __tablename__ = "scenes"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    lesson_id = Column(
        UUID(as_uuid=True),
        ForeignKey("lessons.id", ondelete="CASCADE"),
        nullable=False,
    )
    scene_order = Column(Integer, nullable=False, default=0)
    scene_type = Column(Enum(SceneType), nullable=False)
    title = Column(String(500), nullable=False)
    duration_sec = Column(Float, nullable=False, default=30.0)
    render_strategy = Column(String(255), nullable=False, default="default")
    source_refs_json = Column(JSON, nullable=True, default=list)
    narration_text = Column(Text, nullable=True)
    on_screen_text_json = Column(JSON, nullable=True)
    scene_spec_json = Column(JSON, nullable=True, default=dict)
    status = Column(Enum(SceneStatus), nullable=False, default=SceneStatus.pending)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )

    lesson = relationship("Lesson", back_populates="scenes")
    assets = relationship("SceneAsset", back_populates="scene", cascade="all, delete-orphan")
